x/incentive/keeper/keeper.go

```diff
@@ -39,10 +45,267 @@ func NewKeeper(
+			break
+		}
+	}
+
+	// Find approporiate Incentive Info
+	for _, ii := range params.LpIncentives {
+		if ii.EpochIdentifier == epochIdentifier && timestamp >= ii.StartTime.Unix() {
+			lpIncentive = ii
+		}
+	}
+
+	// return found, stake, lp incentive params
+	return foundIncentive, stakeIncentive, lpIncentive
+}
+
+// Update uncommitted token amount
+// Called back through epoch hook
+func (k Keeper) UpdateUncommittedTokens(ctx sdk.Context, epochIdentifier string) {
+	// Recalculate total committed info
+	k.UpdateTotalCommitmentInfo(ctx, epochIdentifier)
+
+	// Iterate all delegations for the specified delegator
+	// Process to increase uncomitted token amount of Eden & Eden boost
+	k.cmk.IterateCommitments(
+		ctx, epochIdentifier,
+		func(commitments ctypes.Commitments) bool {
+			// Find out active incentive params
+			foundIncentive, stakeIncentive, lpIncentive := k.FindProperIncentiveParm(ctx, epochIdentifier)
+
+			// If we don't have incentive params ready
+			if !foundIncentive {
+				return true
+			}
+
+			// Commitment owner
+			creator := commitments.Creator
+
+			// Calculate delegated amount per delegator
+			delegatedAmt := k.CalculateDelegatedAmount(ctx, creator)
+
+			// Calculate new uncommitted Eden tokens for LP, staker, and Eden token holders
+			new_uncommitted_eden_tokens := k.CalculateNewUncommittedEdenTokens(ctx, delegatedAmt, commitments, stakeIncentive, lpIncentive)
+
+			// Calculate new uncommitted Eden-Boost tokens for staker and Eden token holders
+			new_uncommitted_eden_boost_tokens := k.CalculateNewUncommittedEdenBoostTokens(ctx, delegatedAmt, commitments, epochIdentifier, stakeIncentive, lpIncentive)
+
+			// Update Commitments with new uncommitted token amounts
+			k.UpdateCommitments(ctx, creator, new_uncommitted_eden_tokens, new_uncommitted_eden_boost_tokens)
+
+			return false
+		},
+	)
+}
+
+// Calculate new Eden token amounts based on the given conditions and user's current uncommitted token balance
+func (k Keeper) CalculateNewUncommittedEdenTokens(ctx sdk.Context, delegatedAmt sdk.Int, commitments ctypes.Commitments, stakeIncentive types.IncentiveInfo, lpIncentive types.IncentiveInfo) sdk.Int {
+	// Get LP commitments - Skip for now
+	edenCommittedByLP := sdk.ZeroInt()
+
+	// Get eden commitments and eden boost commitments
+	edenCommitted := sdk.ZeroInt()
+	edenBoostCommitted := sdk.ZeroInt()
+	for _, c := range commitments.CommittedTokens {
+		// Eden committed
+		if c.Denom == types.Eden {
+			edenCommitted = c.Amount
+		}
+
+		// Eden boost committed
+		if c.Denom == types.EdenB {
+			edenBoostCommitted = c.Amount
+		}
+	}
+
+	// compute eden reward based on above and param factors for each
+	totalEdenCommittedByStake := delegatedAmt.Add(edenCommitted).Add(edenBoostCommitted).Add(edenCommittedByLP)
+	stakeShare := k.CalculateTotalShareOfStaking(totalEdenCommittedByStake)
+
+	// Calculate eden amount per epoch
+	edenAmountPerEpoch := stakeIncentive.Amount.Quo(sdk.NewInt(stakeIncentive.NumEpochs))
```

Comment:
@kenta-elys we can probably compute this earlier in the function chain because it will be the same for every commitment object. we don't need to recompute it for every object.

Reply:
We could probably even just change the value stored in the incentive param object to be the amount per epoch so that we don't need to compute it at all.

Reply:
totalEdenCommittedByStake is differs from commitment. It's not the number for entire one, its the number of the committed amount for each commitment.

Reply:
> We could probably even just change the value stored in the incentive param object to be the amount per epoch so that we don't need to compute it at all.

Dream told me about non-linear number for the rewards amount. And it should be implemented inside tokenomic module and incentive module should fetch amount from it.

Reply:
Fixed

Reply:
@kenta-elys 

> here the 'totalEdenCommittedByStake' differs from commitment. It's not the number for entire one, its the number of the committed amount for each commitment.
I was referring to the `edenAmountPerEpoch`, not the `totalEdenCommittedByStake`